File: python/d07.py
from enum import StrEnum

from utils import file_path

with open(file_path(day=7), "r") as f:
    lines = f.readlines()
    splitters = [
        int(line.strip().replace(".", "0").replace("^", "1").replace("S", "1"), 2)
        for line in lines[2::2]
    ]
    width = len(lines[0].strip())


beams = splitters[0]
result = 0
for split in splitters:
    hits = beams & split
    result += hits.bit_count()
    beams -= hits
    beams |= (hits >> 1) | (hits << 1)

print("p1", result)

counts = [int(b) for b in f"{splitters[0]:0{width}b}"]
for split in splitters:
    new_counts = counts.copy()
    for i, count in enumerate(counts):
        collision = split & (1 << i)
        if not collision:
            continue
        new_counts[i] = 0
        new_counts[i - 1] += counts[i]
        new_counts[i + 1] += counts[i]
    counts = new_counts
print("p2", sum(counts))

File: rust/src/bin/d07.rs
use anyhow::{Result, anyhow};
use aoc25::utils::load::load_input;

fn main() -> Result<()> {
    let input = load_input(7)?;

    let (part_one, part_two) = solve(&input).ok_or(anyhow!("no solution"))?;
    println!("part one: {}", part_one);
    println!("part two: {}", part_two);

    Ok(())
}

fn solve(input: &str) -> Option<(u64, u64)> {
    let width = input.lines().next()?.len();
    let mut counts = vec![0u64; width];
    counts[width / 2] = 1;

    let mut p1_result = 0;

    for line in input.lines().skip(2).step_by(2) {
        let mut new_counts = counts.clone();

        for (i, (count, _)) in counts
            .iter()
            .zip(line.chars())
            .enumerate()
            .filter(|(_, (count, c))| **count != 0 && *c == '^')
        {
            p1_result += 1;
            new_counts[i] = 0;
            new_counts[i - 1] += count;
            new_counts[i + 1] += count;
        }

        counts = new_counts;
    }

    Some((p1_result, counts.iter().sum()))
}
